Migration guide for JAX-RS changes (#20659)
keycloak/keycloak: docs/documentation/upgrading/topics/keycloak/changes-22_0_0.adoc

any of these libraries as dependencies of your own providers deployed to Keycloak server, you may also need to copy those `jar` files
explicitly to the Keycloak distribution `providers` directory as well.

= Context and dependency injection no longer enabled to JAX-RS Resources

In order to provide a better runtime and leverage as much as possible the underlying stack,
all injection points for contextual data using the `javax.ws.rs.core.Context` annotation were removed. The expected improvement
in performance involves no longer creating proxies instances multiple times during the request lifecycle, and drastically reducing the amount of reflection code at runtime.

If you are extending one of the following SPIs:

* `PolicySpi`
* `AdminRealmResourceSpi`
* `IdentityProviderSpi`

> * `RealmResourceProvider`

Reviewer: ```suggestion
* `RealmResourceSPI`
```
Should be like this when we're talking about SPIs, right?

Author: Done

> You should review your custom JAX-RS (sub)resources in order to obtain any contextual data from a `org.keycloak.common.util.ResteasyProvider`
> instance as follows:

Reviewer: ```suggestion
You should review your custom JAX-RS (sub)resources in order to obtain any contextual data from a `org.keycloak.common.util.Resteasy` provider as follows:
```
I'd specify the `org.keycloak.common.util.Resteasy` as the developer directly accessing that point. WDYT?

Author: Done

[source,java]
----
KeycloakSession session = org.keycloak.common.util.Resteasy.getContextData(KeycloakSession.class);
----

If you need access to the current request and response objects, you can now obtain their instances directly
from the `KeycloakSession`:

[source,java]
----
@Context
org.jboss.resteasy.spi.HttpRequest request;
@Context
org.jboss.resteasy.spi.HttpResponse response;
----

was replaced by:

[source,java]
----
KeycloakSession session = // obtain the session, which is usually available when creating a custom provider from a factory
KeycloakContext context = session.getContext();

HttpRequest request = context.getHttpRequest();
HttpResponse response = context.getHttpResponse();
----

In case you have no access to a `KeycloakSession` instance when invoking a JAX-RS resource method, you can obtain
contextual data from the JAX-RS runtime as follows:

[source,java]
----
KeycloakSession session = org.keycloak.common.util.Resteasy.getContextData(KeycloakSession.class);
----

Additional contextual data can be obtained from the runtime through the  `KeycloakContext` instance:

[source,java]
----
KeycloakSession session = // obtain the session

> KeycloakContext context = session.getContex();

Reviewer: ```suggestion
KeycloakContext context = session.getContext();
```

Author: Done

MyContextualObject myContextualObject = context.getContextObject(MyContextualObject.class);
----

= Upgrading your custom JAX-RS resources

If you are extending the server's REST APIs through the following SPIs:

* `AdminRealmResourceSpi`
* `IdentityProviderSpi`

> * `RealmResourceProvider`
> * `RealmResourceProvider`

Reviewer: ```suggestion
* `RealmResourceSPI`
```

Should be like this when we're talking about SPIs, right?

Author: Done

> You need to add a `META-INF/beans.xml` to the JAR file where your custom providers are packaged. Otherwise, they are not recognized by the server

Reviewer: ```suggestion
You need to add a `META-INF/beans.xml` file, which can be empty, to the JAR file where your custom providers are packaged. Otherwise, they are not recognized by the server
```
I'd suggest explicitly mentioning that in order to prevent some confusion about what should the beans.xml contain. WDYT?

Author: Done

at runtime.

> You should also remove any type-level `@Path` annotation declaration and mark your methods with this annotation instead.

Reviewer: What does "type-level `@Path`" mean?

Author: At the type, I think it is clear enough ...
